Refactor content distribution functions in Grid

fantasai opened the pull request:
#### 9c0b2b0c72a40ec9e1c7db4f2ff5adcdaa9a7e19
<pre>
Refactor content distribution functions in Grid
<a href="https://bugs.webkit.org/show_bug.cgi?id=266838">https://bugs.webkit.org/show_bug.cgi?id=266838</a>
<a href="https://rdar.apple.com/problem/120064485">rdar://problem/120064485</a>

Reviewed by Sammy Gill.

Improves clarity of content alignment in RenderGrid by:
- Inlines the single-use contentDistributionOffset function into its caller
  to make the logic flow in computeContentPositionAndDistributionOffset more
  straightforward--and to avoid doing a weird thing with an otherwise-
  nonsensical .isValid() that pretends negative numbers are special, but only
  for this one function (because they are otherwise valid in these offsets).
- Reorganizes the logic flow in computeContentPositionAndDistributionOffset
  to start by initializing to zero, and then adjusting where it&apos;s not zero,
  instead of setting things to zero right before every possible return.
- Sets the offsets directly instead of indirectly, now that the control flow
  is simplified.
- Makes it obvious that the offsets are always set, never left uninitialized.

Also fixes a typo in RenderFlexible Box.

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::styleDidChange): Fix typo.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::computeContentPositionAndDistributionOffset): Restructure.
(WebCore::contentDistributionOffset): Deleted (folded into computeContentPositionAndDistributionOffset).
* Source/WebCore/rendering/RenderGrid.h:
(WebCore::ContentAlignmentData::isValid const): Deleted.

Canonical link: <a href="https://commits.webkit.org/272591@main">https://commits.webkit.org/272591@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/18e145702193c718e573edf0323c256916b6c6ee

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32221 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10955 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34036 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34781 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29200 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33071 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13308 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8165 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28765 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9239 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28809 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8042 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8199 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28727 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36126 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29299 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29174 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34323 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8321 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6259 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32186 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9962 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7527 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8954 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8860 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -2241,46 +2241,37 @@ ContentPosition static resolveContentDistributionFallback(ContentDistribution di
 }
 
-static void contentDistributionOffset(ContentAlignmentData& offset, const LayoutUnit& availableFreeSpace, ContentPosition& fallbackPosition, ContentDistribution distribution, unsigned numberOfGridTracks)
+static bool computeDistributionOffset(ContentAlignmentData& offset, const LayoutUnit& availableFreeSpace, ContentPosition& fallbackPosition, ContentDistribution distribution, unsigned numberOfGridTracks)
```
Please return a `{ No, Yes }` enum class here, it's unclear what the bool means when reading this function

fantasai replied:
How is No/Yes clearer than False/True?

I can see adding a comment to make it clearer would be useful, but switching from True/False to Yes/No doesn't make anything clearer to me.

nt1m replied:
You'd be returning and checking against `ContentDistributionIsValid::No` / `ContentDistributionIsValid::Yes` instead of `true` / `false` which is a lot more clear when reading.

The enum class would be defined as: `enum class ContentDistributionIsValid : bool { No, Yes }`

(I'll leave the name of the enum class up to you, the one I gave here was just an example)

nt1m replied:
Fwiw, this is a pretty common pattern in the WebKit codebase: https://searchfox.org/wubkat/search?q=bool+%7B+No%2C+Yes+%7D%3B&path=&case=false&regexp=false when using a boolean would be ambiguous

fantasai replied:
OK, I tried that but it got so annoying to read, I decided to just fold the entire function into its caller, which I had been wanting to do anyway. I think this is easier to follow than splitting out. Lmk if this works for you.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -2293,62 +2251,81 @@ void RenderGrid::computeContentPositionAndDistributionOffset(GridTrackSizingDire
+            offset.positionOffset = availableFreeSpace;
+        return;
     case ContentPosition::Right:
-        ASSERT(isRowAxis);
```
Is removing this assert ok? AFAICT the spec says that left and right work with the justify properties and so we should make sure it's the row axis. I see we kept it in the `ContentPosition::Left` case

fantasai replied:
Agree that's a mistake, fixed.

fantasai committed:
Refactor content distribution functions in Grid
https://bugs.webkit.org/show_bug.cgi?id=266838
rdar://problem/120064485

Reviewed by Sammy Gill.

Improves clarity of content alignment in RenderGrid by:
- Inlines the single-use contentDistributionOffset function into its caller
  to make the logic flow in computeContentPositionAndDistributionOffset more
  straightforward--and to avoid doing a weird thing with an otherwise-
  nonsensical .isValid() that pretends negative numbers are special, but only
  for this one function (because they are otherwise valid in these offsets).
- Reorganizes the logic flow in computeContentPositionAndDistributionOffset
  to start by initializing to zero, and then adjusting where it's not zero,
  instead of setting things to zero right before every possible return.
- Sets the offsets directly instead of indirectly, now that the control flow
  is simplified.
- Makes it obvious that the offsets are always set, never left uninitialized.

Also fixes a typo in RenderFlexible Box.

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::styleDidChange): Fix typo.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::computeContentPositionAndDistributionOffset): Restructure.
(WebCore::contentDistributionOffset): Deleted (folded into computeContentPositionAndDistributionOffset).
* Source/WebCore/rendering/RenderGrid.h:
(WebCore::ContentAlignmentData::isValid const): Deleted.

Canonical link: https://commits.webkit.org/272591@main
